File: MinCostFlow/Graph.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace MinCostFlow
{
    public class Graph
    {
        private const int adjMatrixSize = 20;

        private List<Edge> listOfEdges;
        private List<Vertex> listOfVertices;
        private Dictionary<Vertex, List<Edge>> adjacencyList;
        private double[,] adjacencyMatrix;
        // TODO implement it with secound instance of Graph
        private double[,] residualGraph;
        private double[,] priceGraph;
        private double epsilon;

        public Graph()
        {

            this.listOfEdges = new List<Edge>();
            this.listOfVertices = new List<Vertex>();
            this.adjacencyList = new Dictionary<Vertex, List<Edge>>();
            // TODO make method to take care of the size of the matrix
            this.adjacencyMatrix = new double[adjMatrixSize, adjMatrixSize];
        }

        public void addEdge(Edge newEdge)
        {
            // TODO Validations:
            // 1. Check for parallel arcs
            newEdge = this.updateEdges(newEdge);
            if (this.adjacencyList.ContainsKey(newEdge.From))
            {
                this.adjacencyList[newEdge.From].Add(newEdge);
            }
            else
            {
                this.adjacencyList.Add(newEdge.From, new List<Edge>());
                this.adjacencyList[newEdge.From].Add(newEdge);
            }

            this.updateAdjacencyMatrix();
        }

        private Edge updateEdges(Edge edge)
        {
            edge = this.updateVertices(edge);

            if (!listOfEdges.Contains(edge))
            {
                listOfEdges.Add(edge);
            }

            return edge;
        }

        private Edge updateVertices(Edge edge)
        {
            if (!listOfVertices.Contains(edge.From))
            {
                listOfVertices.Add(edge.From);
                edge.From.Seq = Vertex.NextSeq;
                Vertex.NextSeq++;
            }
            else
            {
                edge.From = this.listOfVertices.Find(x => x.Equals(edge.From));
            }

            if (!listOfVertices.Contains(edge.To))
            {
                listOfVertices.Add(edge.To);
                edge.To.Seq = Vertex.NextSeq;
                Vertex.NextSeq++;
            }
            else
            {
                edge.To = this.listOfVertices.Find(x => x.Equals(edge.To));
            }

            return edge;
        }

        public void removeVertex(Vertex rmVertex)
        {
            this.listOfVertices.Remove(rmVertex);
            var edgesToRemove = new List<Edge>();
            foreach (Edge e in this.listOfEdges)
            {
                if (e.From.Equals(rmVertex) || e.To.Equals(rmVertex))
                {
                    edgesToRemove.Add(e);
                }
            }
            foreach (Edge e in edgesToRemove)
            {
                this.removeEdge(e);
            }

            for (int i = 0; i < adjMatrixSize; i++)
            {
                this.adjacencyMatrix[rmVertex.Seq, i] = 0;
                this.adjacencyMatrix[i, rmVertex.Seq] = 0;
            }

            this.adjacencyList.Remove(rmVertex);
        }

        public bool removeEdge(Edge rmEdge)
        {
            // TODO check what happens if you want to remove an Edge which is not in Graph
            var statusEdges = this.listOfEdges.Remove(rmEdge);
            var statusVertices = this.adjacencyList[rmEdge.From].Remove(rmEdge);
            this.adjacencyMatrix[rmEdge.From.Seq, rmEdge.To.Seq] = 0;

            return statusEdges && statusVertices;
        }

        public double maxFlow(Vertex from, Vertex to)
        {
            residualGraph = new double[adjMatrixSize, adjMatrixSize];
            Vertex start = this.listOfVertices.Find(x => x.Equals(from));
            Vertex end = this.listOfVertices.Find(x => x.Equals(to));

            // TODO make the size more reasonable 
            for (int i = 0; i < adjMatrixSize; i++)
            {
                for (int j = 0; j < adjMatrixSize; j++)
                {
                    this.residualGraph[i, j] = this.adjacencyMatrix[i, j];
                }
            }

            var maxFlow = 0d;
            while (BFS(start, end))
            {
                double minFlow = double.MaxValue;
                // TODO if you use only parenst for BFS you can use only one Vertex
                var path = end;
                while (path.Parents.Count != 0)
                {
                    Vertex u = path.Parents[path.Parents.Count - 1];
                    Vertex v = path;
                    minFlow = Math.Min(minFlow, residualGraph[u.Seq, v.Seq]);

                    path = path.Parents[path.Parents.Count - 1];
                }

                path = end;
                while (path.Parents.Count != 0)
                {
                    Vertex u = path.Parents[path.Parents.Count - 1];
                    Vertex v = path;

                    residualGraph[u.Seq, v.Seq] -= minFlow;
                    residualGraph[v.Seq, u.Seq] += minFlow;

                    path = path.Parents[path.Parents.Count - 1];
                }

                maxFlow += minFlow;
            }

            return maxFlow;
        }

        public Vertex findNegativeCycle(Vertex from)
        {
            Vertex start = this.listOfVertices.Find(x => x.Equals(from));
          
            foreach (Vertex v in this.listOfVertices)
            {
                v.Distance = int.MaxValue;
                v.Parents = new List<Vertex>();
            }
            start.Distance = 0;

            for (int i = 0; i < this.listOfVertices.Count - 1; i++)
            {
                foreach (Edge e in this.listOfEdges)
                {
                    if (e.To.Distance > e.From.Distance + e.Price)
                    {
                        e.To.Distance = e.From.Distance + e.Price;
                        e.To.Parents.Add(e.From);
                    }
                }
            }

            foreach (Edge e in this.listOfEdges)
            {
                if (e.To.Distance > e.From.Distance + e.Price)
                {
                    return e.To;
                }
            }

            return null;
        }

        public Vertex findNegativeCycleInResidualGraph(Vertex from)
        {
            Vertex start = this.listOfVertices.Find(x => x.Equals(from));
            double[] parent = new double[adjMatrixSize];

            parent[start.Seq] = -1;
            foreach (Vertex v in this.listOfVertices)
            {
                v.Distance = int.MaxValue;
                v.Parents = new List<Vertex>();
            }
            start.Distance = 0;

            for (int i = 0; i < this.listOfVertices.Count - 1; i++)
            {
                for (int u = 0; u < adjMatrixSize; u++)
                {
                    for (int v = 0; v < adjMatrixSize; v++)
                    {
                        if (this.residualGraph[u, v] > 0)
                        {
                            //TODO could lead to bugs because I am searching vertex from Seq
                            Vertex uVertex = this.listOfVertices.Find(x => x.Seq == u);
                            Vertex vVertex = this.listOfVertices.Find(x => x.Seq == v);
                            if (vVertex.Distance > uVertex.Distance + this.priceGraph[u, v])
                            {
                                vVertex.Distance = uVertex.Distance + this.priceGraph[u, v];
                                vVertex.Parents.Add(uVertex);
                                vVertex.Parent = uVertex;
                                parent[vVertex.Seq] = uVertex.Seq;
                            }
                        }
                    }
                }
            }

            for (int u = 0; u < adjMatrixSize; u++)
            {
                for (int v = 0; v < adjMatrixSize; v++)
                {
                    if (this.residualGraph[u, v] > 0)
                    {
                        //TODO could lead to bugs because I am searching vertex from Seq
                        Vertex uVertex = this.listOfVertices.Find(x => x.Seq == u);
                        Vertex vVertex = this.listOfVertices.Find(x => x.Seq == v);
                        if (vVertex.Distance > uVertex.Distance + this.priceGraph[u, v])
                        {
                            vVertex.Parents.Add(uVertex);
                            return vVertex;
                        }
                    }
                }
            }

            return null;
        }

        public int minCostFlowCycleCancel(Vertex from, Vertex to, int cargo)
        {
            Vertex start = this.listOfVertices.Find(x => x.Equals(from));
            Vertex end = this.listOfVertices.Find(x => x.Equals(to));
            Vertex source = new Vertex("s");
            Vertex dest = new Vertex("t");

            this.addEdge(new Edge(source, start, cargo, 0));
            this.addEdge(new Edge(end, dest, cargo, 0));
            //TODO make the flow to be int
            int maxFlow = (int)this.maxFlow(source, dest);

            if (maxFlow < cargo)
            {
                //TODO make custom exception
                throw new ApplicationException("There is no feasible solution for this sypply: " + cargo);
            }
            else
            {
                this.establishFeasibleFLow(source, dest);
                this.generatePriceGraph();
            }

            Vertex vertexInLoop = this.findNegativeCycleInResidualGraph(start);
            while (vertexInLoop != null)
            {
                Vertex parent = vertexInLoop.Parents[vertexInLoop.Parents.Count - 1];
                List<Vertex> cycle = new List<Vertex>();
                //cycle.Add(vertexInLoop);
                while (!cycle.Contains(parent)/*parent != vertexInLoop*/)
                {
                    cycle.Add(parent);
                    parent = parent.Parents[parent.Parents.Count - 1];
                }

                List<double> rFlows = new List<double>();
                for (int i = 0; i < cycle.Count; i++)
                {
                    if (i == cycle.Count - 1)
                    {
                        rFlows.Add(this.residualGraph[cycle[0].Seq, cycle[cycle.Count - 1].Seq]);
                    }
                    else
                    {
                        rFlows.Add(this.residualGraph[cycle[i + 1].Seq, cycle[i].Seq]);
                    }
                }

                double minRFlow = rFlows.Min();

                for (int i = 0; i < cycle.Count; i++)
                {
                    if (i == cycle.Count - 1)
                    {
                        this.residualGraph[cycle[0].Seq, cycle[cycle.Count - 1].Seq] -= minRFlow;
                        this.residualGraph[cycle[cycle.Count - 1].Seq, cycle[0].Seq] += minRFlow;
                    }
                    else
                    {
                        this.residualGraph[cycle[i + 1].Seq, cycle[i].Seq] -= minRFlow;
                        this.residualGraph[cycle[i].Seq, cycle[i + 1].Seq] += minRFlow;
                    }
                }

                vertexInLoop = this.findNegativeCycleInResidualGraph(start);
            }

            return 0;
        }

        public int minCostFlowCostScaling(Vertex from, Vertex to, int cargo)
        {
            Vertex start = this.listOfVertices.Find(x => x.Equals(from));
            Vertex end = this.listOfVertices.Find(x => x.Equals(to));

            start.NodeImbalance = cargo;
            end.NodeImbalance = -cargo;

            this.epsilon = listOfEdges.Max(x => x.Price);
            foreach (Vertex v in this.listOfVertices)
            {
                v.Pi = 0;
            }

            Vertex source = new Vertex("s");
            Vertex dest = new Vertex("t");
            this.addEdge(new Edge(source, start, cargo, 0));
            this.addEdge(new Edge(end, dest, cargo, 0));
            //TODO make the flow to be int
            int maxFlow = (int)this.maxFlow(source, dest);

            if (maxFlow < cargo)
            {
                //TODO make custom exception
                throw new ApplicationException("There is no feasible solution for this sypply: " + cargo);
            }
            else
            {
                // TODO probably should remove the flow form the residual network
                this.establishFeasibleFLowCostScaling(source, dest);
                this.generatePriceGraph();
            }

            while (this.epsilon >= (Double)1 / this.listOfEdges.Count)
            {
                this.improveApproxipation();
                this.epsilon = this.epsilon / 2;

                //if(this.listOfVertices.Find(x => x.NodeImbalance > 0) == null)
                //{
                //    break;
                //}
            }

            return 0;
        }

        private void improveApproxipation()
        {
            foreach (Edge e in this.listOfEdges)
            {
                if (e.Price - e.From.Pi + e.To.Pi > 0)
                {
                    this.residualGraph[e.From.Seq, e.To.Seq] += this.residualGraph[e.To.Seq, e.From.Seq];
                    this.residualGraph[e.To.Seq, e.From.Seq] = 0;
                }
                //else if (e.Price - e.From.Pi + e.To.Pi < 0)
                //{
                //    this.residualGraph[e.To.Seq, e.From.Seq] += this.residualGraph[e.From.Seq, e.To.Seq];
                //    this.residualGraph[e.From.Seq, e.To.Seq] = 0;
                //}
            }

            ////TODO make it in method
            //foreach (Vertex v in this.listOfVertices)
            //{
            //    v.NodeImbalance = v.NodeBalance;

            //    //TODO check if your residual network representation is correct because it could be wrong
            //    for (int i = 0; i < adjMatrixSize; i++)
            //    {
            //        v.NodeImbalance += this.residualGraph[i, v.Seq];
            //        v.NodeImbalance -= this.residualGraph[v.Seq, i];
            //    }
            //}

            Vertex activeVertex = this.listOfVertices.Find(x => x.NodeImbalance > 0);
            while (activeVertex != null)
            {
                var check = 0;
                for (int i = 0; i < adjMatrixSize; i++)
                {
                    if (this.residualGraph[activeVertex.Seq, i] == 0)
                    {
                        continue;
                    }

                    Vertex jVertex = this.listOfVertices.Find(x => x.Seq == i);
                    if ((-1 * this.epsilon / 2) <= (this.priceGraph[activeVertex.Seq, jVertex.Seq] - activeVertex.Pi + jVertex.Pi) &&
                        (this.priceGraph[activeVertex.Seq, jVertex.Seq] - activeVertex.Pi + jVertex.Pi) < 0)
                    {
                        // TODO make the var types correct the flow should be probably Integer
                        Double flow = Math.Min(activeVertex.NodeImbalance, this.residualGraph[activeVertex.Seq, jVertex.Seq]);

                        this.residualGraph[jVertex.Seq, activeVertex.Seq] = this.residualGraph[jVertex.Seq, activeVertex.Seq] + flow;
                        this.residualGraph[activeVertex.Seq, jVertex.Seq] = this.residualGraph[activeVertex.Seq, jVertex.Seq] - flow;

                        activeVertex.NodeImbalance -= flow;
                        jVertex.NodeImbalance += flow;

                        check = 0;
                        // TODO TEST CASE ALG not sure if we should break the cycle
                        break;
                    }
                    else
                    {
                        check++;
                    }

                }

                if(check > 0)
                {
                    // TODO make the whole thing to work with floats or just all prices * 2 in oder preserve epsilon to be whole number
                    activeVertex.Pi += + (int)this.epsilon / 2;
                }


                if (activeVertex.NodeImbalance <= 0)
                {
                    // TODO if the activeVertex stays active should be pickeged again?
                    activeVertex = this.listOfVertices.Find(x => x.NodeImbalance > 0);
                }
            }
        }

        public void generatePriceGraph()
        {
            //TODO Clean it also when you are deleteing vertices or edges
            priceGraph = new double[adjMatrixSize, adjMatrixSize];

            foreach (Edge e in this.listOfEdges)
            {
                priceGraph[e.From.Seq, e.To.Seq] = e.Price;
                priceGraph[e.To.Seq, e.From.Seq] = -e.Price;
            }
        }

        private void establishFeasibleFLowCostScaling(Vertex source, Vertex dest)
        {
            this.removeVertex(source);
            this.removeVertex(dest);

            for (int i = 0; i < adjMatrixSize; i++)
            {
                // The the indexes are reversed because in the residual network there is no edge "s" -> start and end -> "t"
                // because we add artificial edges with the desired capacity of the supply and demand.
                // That is why these edges are with full capacity and we have only the reversed edges in the Residual network
                this.residualGraph[i, source.Seq] = 0;
                this.residualGraph[dest.Seq, i] = 0;
            }
        }

        private void establishFeasibleFLow(Vertex source, Vertex dest)
        {
            this.removeVertex(source);
            this.removeVertex(dest);

            for (int i = 0; i < adjMatrixSize; i++)
            {
                // The the indexes are reversed because in the residual network there is no edge "s" -> start and end -> "t"
                // because we add artificial edges with the desired capacity of the supply and demand.
                // That is why these edges are with full capacity and we have only the reversed edges in the Residual network
                this.residualGraph[i, source.Seq] = 0;
                this.residualGraph[dest.Seq, i] = 0;
            }
        }

        private bool BFS(Vertex from, Vertex to)
        {
            var verticesCnt = this.listOfVertices.Count;
            foreach (Vertex vertex in this.listOfVertices)
            {
                vertex.IsVisited = false;
            }

            Queue<Vertex> queue = new Queue<Vertex>();
            queue.Enqueue(from);
            from.IsVisited = true;

            while (queue.Count != 0)
            {
                Vertex curr = queue.Dequeue();

                foreach (Vertex v in this.listOfVertices)
                {
                    if (v.IsVisited == false && this.residualGraph[curr.Seq, v.Seq] > 0)
                    {
                        queue.Enqueue(v);
                        v.addParent(curr);
                        v.IsVisited = true;
                    }
                }
            }

            return (to.IsVisited == true);
        }

        private void updateAdjacencyMatrix()
        {
            foreach (Vertex v in this.adjacencyList.Keys)
            {
                foreach (Edge e in this.adjacencyList[v])
                {
                    if (adjMatrixSize > e.From.Seq && adjMatrixSize > e.To.Seq)
                    {
                        adjacencyMatrix[e.From.Seq, e.To.Seq] = e.Capacity;
                    }
                    else
                    {
                        this.updatadjMatrixSize();
                    }
                }
            }
        }

        public void printGraphMinCostFlow()
        {
            double minCostFlow = 0;
            foreach (Edge e in this.listOfEdges)
            {
                minCostFlow += this.residualGraph[e.To.Seq, e.From.Seq] * this.priceGraph[e.From.Seq, e.To.Seq];
                Console.WriteLine(e.From + " -> " + e.To + " - Flow: " + this.residualGraph[e.To.Seq, e.From.Seq] + " / Price: " + this.residualGraph[e.To.Seq, e.From.Seq] * this.priceGraph[e.From.Seq, e.To.Seq]);
            }
            Console.WriteLine("The min const flow is: " + minCostFlow);
        }

        public void printGraphMaxFlow()
        {
            foreach (Edge e in this.listOfEdges)
            {
                Console.WriteLine(e.From + " -> " + e.To + " = " + this.residualGraph[e.To.Seq, e.From.Seq]);
            }
        }

        public void printVertexSeq()
        {
            foreach (Vertex v in this.listOfVertices)
            {
                Console.WriteLine(v + " -> " + v.Seq);
            }
        }

        // TODO implement the method
        private void updatadjMatrixSize()
        {
            // Probably you need to rebuild Seq of vertex
            throw new NotImplementedException();
        }

        //TODO: What happens if the goal destination has an outgoing arc
    }
}

File: opticost/src/main/java/com/project/opticost/algorithm/Graph.java
package com.project.opticost.algorithm;

import com.project.opticost.utils.exceptions.NoFeasibleSolutionException;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.*;

public class Graph {
    private List<Edge> listOfEdges;
    private List<Vertex> listOfVertices;
    private Map<Vertex, List<Edge>> adjacencyList;
    private Map<Vertex, List<ResidualEdge>> residualGraph;
    private BigDecimal minCostFlow;

    public Graph() {
        this.listOfEdges = new ArrayList<>();
        this.listOfVertices = new ArrayList<>();
        this.adjacencyList = new HashMap<>();
        this.residualGraph = new HashMap<>();
    }

    public BigDecimal getMinCostFlow() {
        return minCostFlow;
    }

    public void addEdge(Edge newEdge) {
        newEdge = this.updateEdges(newEdge);
        if (this.adjacencyList.containsKey(newEdge.getFrom())) {
            this.adjacencyList.get(newEdge.getFrom()).add(newEdge);
        } else {
            this.adjacencyList.put(newEdge.getFrom(), new ArrayList<>());
            this.adjacencyList.get(newEdge.getFrom()).add(newEdge);
        }
    }

    private Edge updateEdges(Edge edge) {
        edge = this.updateVertices(edge);

        if (!listOfEdges.contains(edge)) {
            listOfEdges.add(edge);
        }

        return edge;
    }

    private Edge updateVertices(Edge edge) {
        if (!listOfVertices.contains(edge.getFrom())) {
            int seq = Vertex.getNextSeq();
            listOfVertices.add(edge.getFrom());
            edge.getFrom().setSeq(seq);
            Vertex.setNextSeq(seq + 1);
        } else {
            edge.setFrom(this.listOfVertices.stream()
                    .filter(x -> x.equals(edge.getFrom()))
                    .findAny().orElse(null));
        }

        if (!listOfVertices.contains(edge.getTo())) {
            int seq = Vertex.getNextSeq();
            listOfVertices.add(edge.getTo());
            edge.getTo().setSeq(seq);
            Vertex.setNextSeq(seq + 1);
        } else {
            edge.setTo(this.listOfVertices.stream()
                    .filter(x -> x.equals(edge.getTo()))
                    .findAny().orElse(null));
        }

        return edge;
    }

    public void removeVertex(Vertex rmVertex) {
        this.listOfVertices.remove(rmVertex);
        List<Edge> edgesToRemove = new ArrayList<>();
        for (Edge e : this.listOfEdges) {
            if (e.getFrom().equals(rmVertex) || e.getTo().equals(rmVertex)) {
                edgesToRemove.add(e);
            }
        }

        for (Edge e : edgesToRemove) {
            this.removeEdge(e);
        }
        this.adjacencyList.remove(rmVertex);
    }

    public boolean removeEdge(Edge rmEdge) {
        boolean statusEdges = this.listOfEdges.remove(rmEdge);
        boolean statusVertices = this.adjacencyList.get(rmEdge.getFrom()).remove(rmEdge);

        return statusEdges && statusVertices;
    }

    public Integer maxFlow(Vertex from, Vertex to) {
        Vertex start = this.listOfVertices.stream().filter(x -> x.equals(from)).findAny().orElse(null);
        Vertex end = this.listOfVertices.stream().filter(x -> x.equals(to)).findAny().orElse(null);

        buildResidualGraph();

        Integer maxFlow = 0;
        while (BFS(start, end)) {
            Integer minFlow = Integer.MAX_VALUE;
            Vertex path = end;
            List<ResidualEdge> pathEdges = new ArrayList<>();
            while (path.getParents().size() != 0) {
                Vertex u = path.getParents().get(path.getParents().size() - 1);
                Vertex v = path;

                ResidualEdge edge = residualGraph.get(u).stream()
                        .filter(x -> x.getTo().equals(v) && x.getFlow() > 0)
                        .findAny().orElseThrow(NoSuchElementException::new);

                minFlow = Math.min(minFlow, edge.getFlow());

                path = path.getParents().get(path.getParents().size() - 1);
                pathEdges.add(edge);
            }

            for (ResidualEdge pathEdge : pathEdges) {
                pathEdge.setFlow(pathEdge.getFlow() - minFlow);
                pathEdge.getMirrorEdge().setFlow(pathEdge.getMirrorEdge().getFlow() + minFlow);
            }

            maxFlow += minFlow;
        }

        return maxFlow;
    }

    public Vertex findNegativeCycleInResidualGraph(Vertex from) {
        Vertex start = this.listOfVertices.stream().filter(x -> x.equals(from)).findAny().orElse(null);
        for (Vertex v : this.listOfVertices) {
            v.setDistance((double) Integer.MAX_VALUE);
            v.setParents(new ArrayList<>());
        }
        start.setDistance(0d);
        for (int i = 0; i < this.listOfVertices.size() - 1; i++) {
            for (Map.Entry<Vertex, List<ResidualEdge>> entry : this.residualGraph.entrySet()) {
                Vertex uVertex = this.listOfVertices.stream().filter(x -> x.equals(entry.getKey())).findAny().orElse(null);
                for (ResidualEdge e : entry.getValue()) {
                    if (e.getFlow() > 0) {
                        Vertex vVertex = this.listOfVertices.stream().filter(x -> x.equals(e.getTo())).findAny().orElse(null);
                        if (vVertex.getDistance() > uVertex.getDistance() + e.getPrice().doubleValue()) {
                            vVertex.setDistance(uVertex.getDistance() + e.getPrice().doubleValue());
                            vVertex.setParent(new Parent(uVertex, e));
                        }
                    }
                }
            }
        }

        for (Map.Entry<Vertex, List<ResidualEdge>> entry : this.residualGraph.entrySet()) {
            Vertex uVertex = this.listOfVertices.stream().filter(x -> x.equals(entry.getKey())).findAny().orElse(null);
            for (ResidualEdge e : entry.getValue()) {
                if (e.getFlow() > 0) {
                    Vertex vVertex = this.listOfVertices.stream().filter(x -> x.equals(e.getTo())).findAny().orElse(null);
                    if (vVertex.getDistance() > uVertex.getDistance() + e.getPrice().doubleValue()) {
                        vVertex.setParent(new Parent(uVertex, e));
                        return vVertex;
                    }
                }
            }
        }
        return null;
    }

    public void minCostFlowCycleCancel(Vertex from, Vertex to, Integer cargo) throws Exception {
        Vertex start = this.listOfVertices.stream().filter(x -> x.equals(from)).findAny().orElse(null);
        Vertex end = this.listOfVertices.stream().filter(x -> x.equals(to)).findAny().orElse(null);

        Vertex source = new Vertex("s");
        Vertex dest = new Vertex("t");

        this.addEdge(new Edge(source, start, cargo, BigDecimal.valueOf(0)));
        this.addEdge(new Edge(end, dest, cargo, BigDecimal.valueOf(0)));
        Integer maxFlow = this.maxFlow(source, dest);

        if (maxFlow < cargo) {
            throw new NoFeasibleSolutionException("There is no feasible solution for the supply: " + cargo + "." +
                    "The max flow between the two cities is: " + maxFlow);
        } else {
            this.establishFeasibleFLow(source, dest);
        }

        Vertex vertexInLoopPath = this.findNegativeCycleInResidualGraph(start);
        while (vertexInLoopPath != null) {
            Vertex parent = vertexInLoopPath.getParent().getVertex();
            List<Vertex> cycle = new ArrayList<>();
            cycle.add(vertexInLoopPath);
            while (!cycle.contains(parent)) {
                cycle.add(parent);
                parent = parent.getParent().getVertex();
            }
            cycle = cycle.subList(cycle.indexOf(parent), cycle.size());

            List<ResidualEdge> singleEdges = new ArrayList<>();
            for (Vertex v : cycle) {
                singleEdges.add(v.getParent().getResidualEdge());
            }

            Integer minRFlow = singleEdges.stream().min(Comparator.comparing(ResidualEdge::getFlow)).get().getFlow();
            for (ResidualEdge singleEdge : singleEdges) {
                Integer flowMinus = singleEdge.getFlow();
                singleEdge.setFlow(flowMinus - minRFlow);

                Integer flowPlus = singleEdge.getMirrorEdge().getFlow();
                singleEdge.getMirrorEdge().setFlow(flowPlus + minRFlow);
            }
            vertexInLoopPath = this.findNegativeCycleInResidualGraph(start);
        }
    }

    private void establishFeasibleFLow(Vertex source, Vertex dest) {
        this.removeVertex(source);
        this.removeVertex(dest);

        this.residualGraph.remove(source);
        this.residualGraph.remove(dest);

        for (Map.Entry<Vertex, List<ResidualEdge>> entry : this.residualGraph.entrySet()) {
            List<ResidualEdge> toRemove = new ArrayList<>();
            for (ResidualEdge e : entry.getValue()) {
                if (e.getFrom().equals(source) ||
                        e.getTo().equals(source) ||
                        e.getFrom().equals(dest) ||
                        e.getTo().equals(dest)) {
                    toRemove.add(e);
                }
            }

            for (ResidualEdge e : toRemove) {
                entry.getValue().remove(e);
            }
        }
    }

    private boolean BFS(Vertex from, Vertex to) {
        for (Vertex vertex : this.listOfVertices) {
            vertex.setVisited(false);
            vertex.setParents(new ArrayList<>());
        }

        Queue<Vertex> queue = new ArrayDeque<>();
        ((ArrayDeque<Vertex>) queue).addLast(from);
        from.setVisited(true);

        while (queue.size() != 0) {
            Vertex curr = ((ArrayDeque<Vertex>) queue).removeFirst();
            for (Vertex v : this.listOfVertices) {
                // Creates fake flow for not connected vertices based on the BFS implementation
                double flow = residualGraph.get(curr).stream()
                        .filter(x -> x.getTo().equals(v) && x.getFlow() > 0)
                        .findAny().orElse(new ResidualEdge(curr, v, -1, BigDecimal.ZERO, false)).getFlow();

                if (v.isVisited() == false && flow > 0) {
                    ((ArrayDeque<Vertex>) queue).addLast(v);
                    v.addParent(curr);
                    v.setVisited(true);
                }
            }
        }

        return (to.isVisited() == true);
    }

    private void buildResidualGraph() {
        for (Map.Entry<Vertex, List<Edge>> entry : this.adjacencyList.entrySet()) {
            if (!residualGraph.containsKey(entry.getKey())) {
                residualGraph.put(entry.getKey(), new ArrayList<>());
            }
            for (Edge e : entry.getValue()) {
                ResidualEdge fromTo = new ResidualEdge(e.getFrom(), e.getTo(), e.getCapacity(), e.getPrice(), false);
                ResidualEdge toFrom = new ResidualEdge(e.getTo(), e.getFrom(), 0, e.getPrice().negate(), true);

                fromTo.setMirrorEdge(toFrom);
                toFrom.setMirrorEdge(fromTo);

                residualGraph.get(entry.getKey()).add(fromTo);
                if (!residualGraph.containsKey(e.getTo())) {
                    residualGraph.put(e.getTo(), new ArrayList<>());
                    residualGraph.get(e.getTo()).add(toFrom);
                } else {
                    residualGraph.get(e.getTo()).add(toFrom);
                }
            }
        }
    }

    public List<ResultEdge> getResult() {
        BigDecimal minCostFlow = BigDecimal.valueOf(0);
        List<ResultEdge> result = new ArrayList<>();
        for (Map.Entry<Vertex, List<ResidualEdge>> entry : this.residualGraph.entrySet()) {
            for (ResidualEdge e : entry.getValue()) {
                if (e.isResult()) {
                    BigDecimal price = BigDecimal.valueOf(e.getFlow() * e.getPrice().negate().doubleValue()).setScale(2, RoundingMode.HALF_UP);
                    minCostFlow = minCostFlow.add(price);
                    result.add(new ResultEdge(e.getTo().getName(), e.getFrom().getName(), price, e.getFlow()));
                }
            }
        }

        this.minCostFlow = minCostFlow;
        return result;
    }

    public void printGraphMinCostFlow() {
        BigDecimal minCostFlow = BigDecimal.valueOf(0);
        for (Map.Entry<Vertex, List<ResidualEdge>> entry : this.residualGraph.entrySet()) {
            for (ResidualEdge e : entry.getValue()) {
                if (e.isResult()) {
                    BigDecimal price = BigDecimal.valueOf(e.getFlow() * e.getPrice().negate().doubleValue()).setScale(2, RoundingMode.HALF_UP);
                    minCostFlow = minCostFlow.add(price);

                    System.out.println(e.getTo() + " -> " + e.getFrom() + " - Flow: " + e.getFlow() +
                            " / Price: " + price.doubleValue());
                }

            }
        }
        System.out.println("The min const flow is: " + minCostFlow);
    }
}
